File: project_euler_unified/problem_362/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Problem 362: Squarefree Factors
 *
 * f(n) = largest squarefree divisor of n.
 * Compute sum of f(n) for n = 1 to 10^14.
 *
 * Approach: Write n = a^2 * b (b squarefree), then f(n) = b.
 * Use Mobius function for squarefree sum computation.
 *
 * Answer: 457895958010
 */

typedef long long ll;
typedef __int128 lll;

const ll N = 100000000000000LL; // 10^14

// Sieve smallest prime factor up to limit
vector<int> mu_sieve;
vector<int> mobius;

void compute_mobius(int limit) {
    mobius.assign(limit + 1, 0);
    mobius[1] = 1;
    vector<int> smallest_prime(limit + 1, 0);
    vector<int> primes;

    for (int i = 2; i <= limit; i++) {
        if (smallest_prime[i] == 0) {
            smallest_prime[i] = i;
            primes.push_back(i);
            mobius[i] = -1;
        }
        for (int j = 0; j < (int)primes.size() && primes[j] <= smallest_prime[i] && (ll)i * primes[j] <= limit; j++) {
            smallest_prime[i * primes[j]] = primes[j];
            if (i % primes[j] == 0) {
                mobius[i * primes[j]] = 0;
            } else {
                mobius[i * primes[j]] = -mobius[i];
            }
        }
    }
}

// T(m) = m*(m+1)/2, using 128-bit to avoid overflow
lll T(ll m) {
    return (lll)m * (m + 1) / 2;
}

// Sum of squarefree numbers up to x
lll squarefree_sum(ll x) {
    ll sq = (ll)sqrt((double)x);
    while ((sq + 1) * (sq + 1) <= x) sq++;
    while (sq * sq > x) sq--;

    lll result = 0;
    for (ll d = 1; d <= sq; d++) {
        if (mobius[d] == 0) continue;
        ll q = x / (d * d);
        result += (lll)mobius[d] * d * d * T(q);
    }
    return result;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    ll sqrtN = (ll)sqrt((double)N);
    while ((sqrtN + 1) * (sqrtN + 1) <= N) sqrtN++;
    while (sqrtN * sqrtN > N) sqrtN--;

    // We need mobius values up to sqrt(N/1) = sqrt(N) ~ 10^7
    int mobius_limit = (int)sqrtN + 1;
    compute_mobius(mobius_limit);

    lll answer = 0;
    for (ll a = 1; a <= sqrtN; a++) {
        ll limit = N / (a * a);
        answer += squarefree_sum(limit);
    }

    cout << (ll)answer << endl;
    // Expected: 457895958010

    return 0;
}

File: project_euler_unified/problem_362/solution.py
"""
Problem 362: Squarefree Factors

f(n) = largest squarefree divisor of n.
Compute sum of f(n) for n = 1 to 10^14.

Approach:
- Write n = a^2 * b where b is squarefree, then f(n) = b
- Sum becomes: sum over a of (sum of squarefree b <= N/a^2)
- Use Mobius function for squarefree sum computation

Answer: 457895958010
"""

import math

def compute_mobius(limit):
    """Compute Mobius function via sieve up to limit."""
    mu = [0] * (limit + 1)
    mu[1] = 1
    is_prime = [True] * (limit + 1)
    primes = []

    for i in range(2, limit + 1):
        if is_prime[i]:
            primes.append(i)
            mu[i] = -1
        for p in primes:
            if i * p > limit:
                break
            is_prime[i * p] = False
            if i % p == 0:
                mu[i * p] = 0
                break
            else:
                mu[i * p] = -mu[i]

    return mu

def T(m):
    """Triangular number: m*(m+1)//2."""
    return m * (m + 1) // 2

def squarefree_sum(x, mu):
    """Sum of all squarefree numbers from 1 to x."""
    sq = int(math.isqrt(x))
    result = 0
    for d in range(1, sq + 1):
        if mu[d] == 0:
            continue
        q = x // (d * d)
        result += mu[d] * d * d * T(q)
    return result

def solve():
    N = 10**14
    sqrtN = int(math.isqrt(N))

    # Need mobius values up to sqrt(N) ~ 10^7
    mu = compute_mobius(sqrtN + 1)

    answer = 0
    for a in range(1, sqrtN + 1):
        limit = N // (a * a)
        if limit == 0:
            break
        answer += squarefree_sum(limit, mu)

    return answer

if __name__ == "__main__":
    answer = solve()
assert answer == 457895958010
print(answer)
    # Expected: 457895958010
